data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt

```diff
@@ -0,0 +1,200 @@
+import org.oppia.android.util.data.DataProvidersInjectorProvider
+import org.robolectric.Shadows
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Qualifier
+import javax.inject.Singleton
+
+/** Tests for [RemoteAuthNetworkInterceptor] */
+@RunWith(AndroidJUnit4::class)
+@Config(application = RemoteAuthNetworkInterceptorTest.TestApplication::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+class RemoteAuthNetworkInterceptorTest {
+
+  @Inject
+  lateinit var remoteAuthNetworkInterceptor: RemoteAuthNetworkInterceptor
+
+  @Inject
+  lateinit var context: Context
+
+  private lateinit var retrofit: Retrofit
+
+  lateinit var topicService: TopicService
+
+  private val testVersionName = "1.0"
+
+  private val testVersionCode = 1
+
+  private val topicName = "Topic1"
+
+  private val mockWebServer = MockWebServer()
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+    setUpApplicationForContext()
+    setUpRetrofit()
+  }
+
+  private fun setUpApplicationForContext() {
+    val packageManager = Shadows.shadowOf(context.packageManager)
+    val applicationInfo =
+      ApplicationInfoBuilder.newBuilder()
+        .setPackageName(context.packageName)
+        .build()
+    val packageInfo =
+      PackageInfoBuilder.newBuilder()
+        .setPackageName(context.packageName)
+        .setApplicationInfo(applicationInfo)
+        .build()
+    packageInfo.versionName = testVersionName
+    packageInfo.versionCode = testVersionCode
+    packageManager.installPackage(packageInfo)
+  }
+
+  @Test
+  fun testNetworkInterceptor_withoutHeaders_addsCorrectHeaders() {
+    mockWebServer.enqueue(MockResponse().setBody("{}"))
+    val call = topicService.getTopicByName(topicName)
+    val serviceRequest = call.request()
+    assertThat(serviceRequest.header("api_key")).isNull()
+    assertThat(serviceRequest.header("app_package_name")).isNull()
+    assertThat(serviceRequest.header("app_version_name")).isNull()
+    assertThat(serviceRequest.header("app_version_code")).isNull()
+
+    call.execute()
+    val interceptedRequest = mockWebServer.takeRequest(timeout = 8, unit = TimeUnit.SECONDS)
+    verifyRequestHeaders(interceptedRequest!!.headers)
+  }
+
+  @Test
+  fun testNetworkInterceptor_withHeaders_setsCorrectHeaders() {
+    mockWebServer.enqueue(MockResponse().setBody("{}"))
+    val call = topicService.getTopicByName(topicName)
+
+    val serviceRequest = topicService.getTopicByName(topicName).request()
+    val request = serviceRequest.newBuilder()
+      .addHeader("api_key", "wrong_api_key")
+      .addHeader("app_package_name", "wrong_package_name")
+      .addHeader("app_version_name", "wrong_version_name")
+      .addHeader("app_version_code", "wrong_version_code")
+      .build()
```

Comment:
@BenHenning I realized that this test doesn't actually use the request with the incorrectly set headers. I don't believe MockWebServer enables us to enqueue a specific request, nor do I know how to pass a request directly into the service (unless we want to mock the service differently). Do you have any suggestions on how we could test this case?

Reply:
Per offline discussion I think we landed on using Okhttp directly (& per the latest code)--can this be resolved now?

Reply:
Ah yup, thanks for following-up on this